harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -271,75 +498,80 @@ def cleanup_iterator() -> None:
-                    action = "training"
-                    metrics = self._train_for_step(
-                        w.step_id,
-                        w.num_batches,
-                        w.total_batches_processed,
-                    )
-                    response = {
-                        "metrics": metrics,
-                        "stop_requested": self.context.get_stop_requested(),
-                    }  # type: workload.Response
-                    metrics = self.context.distributed.broadcast(metrics)
-                    for callback in self.callbacks.values():
-                        callback.on_training_workload_end(
-                            avg_metrics=metrics["avg_metrics"],
-                            batch_metrics=metrics["batch_metrics"],
-                        )
-                    if (
-                        self.context.distributed.size > 1
-                        and not self.context._average_training_metrics
-                    ):
-                        warnings.warn(
-                            "Only the chief worker's training metrics are being reported, due "
-                            "to setting average_training_metrics to False.",
-                            UserWarning,
-                        )
-                elif w.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
-                    action = "validation"
-                    response = {
-                        "metrics": self._compute_validation_metrics(),
-                        "stop_requested": self.context.get_stop_requested(),
-                    }
-                elif w.kind == workload.Workload.Kind.CHECKPOINT_MODEL:
-                    action = "checkpointing"
-                    uuid = ""
-                    if self.is_chief:
-                        metadata = {
-                            "determined_version": det.__version__,
-                            "steps_completed": self.steps_completed,
-                            "framework": f"torch-{torch.__version__}",
-                            "format": "pickle",
-                        }
-                        with self.context._core.checkpoint.store_path(metadata) as (
-                            path,
-                            storage_id,
-                        ):
-                            self._save(path)
-                            uuid = storage_id
-                        response = {"uuid": storage_id}
-                    else:
-                        response = {}
-                    uuid = self.context.distributed.broadcast(uuid)
-                    for callback in self.callbacks.values():
-                        callback.on_checkpoint_upload_end(uuid=uuid)
+                min_checkpoint_period = self.env.experiment_config.get_min_checkpoint_period()
+                min_validation_period = self.env.experiment_config.get_min_validation_period()
+                try:
```

Comment:
at this level of indentation, I'd recommend everything inside the `try` clause be a separate function.  That makes `run` primarily about setting up context managers, and `_run` or whatever you want to call it primarily business logic.  It's a nice separation.

Reply:
done